### PR TITLE
Add support for literal argument default values in Python interoperability

### LINKED ISSUE #10667: Handle default values for exported functions in python library compilation
I believe we lose the idea of default values after function resolution, meaning that codegen is too late to output them.  Standard C does not support default values, so such information can't live in the header or .pxd file.  However, Python does allow them, and you can specify the value in the .pyx's function declaration, e.g.:

```cython
from foo cimport functionName as chpl_functionName

def functionName (arg1 = 0):
  chpl_functionName(arg1)
```

Imo, we should make an effort to support this case.  Our AI workflow implementation also relies on it (made possible today by writing the .pyx files ourselves).

### PR DESCRIPTION
**Pull request by lydia-duncan**
C does not support default values for arguments, but Python does.  Add support for the
default values when interoperating with Python, and a warning message when the code
is compiled for interoperating with C.  Currently only handles when literals are used as
the default value, but we anticipate being able to support cases like computations,
function calls, referring to previous arguments, and referring to exported global
variables in the future.

Resolves #10667 

Details:
Default values for arguments are no longer relevant to the AST after function
resolution and get cleaned up between there and code generation.  So when
resolving exported functions, save the value into a map from the argument to
a string.  Then, at code generation, output the default value into the function
header in the .pyx file.

Also made a minor change to the param cname support for exported functions
so that they wouldn't trigger my analysis (since they had been removed from the
tree already anyways).

Generates a warning that the default value has been dropped on the floor for:
- C, on any exported function
- Python when the default value is not a literal like `4`, `"blah"`, etc.

Adds 6 new tests, covering:
- The warning message for C library compilation
- Ensures that even with the warning, the library function can still be called when
given a value for the argument with a default
- support for default literal values for `c_string`s, `bool`s, `int`s, `real`s, `uint`s
- a future for the default value of an argument of type `complex` (this test also covers
when the default value is a computation of some sort instead of just a recognizable
literal)
- a future for when the default value references an exported global variable
- a future for when the default value references a prior argument

Testing:
- Passed full paratest with futures
- Passed fresh checkout on python testing machine

### CODE REVIEW

File: compiler/resolution/resolveFunction.cpp [hunk 1]
```diff
@@ -143,6 +149,91 @@ static void resolveFormals(FnSymbol* fn) {
...
+          }
+          exportedDefaultValues[formal] = ss.str();
+          break;
+        }
+
+        case NUM_KIND_COMPLEX: {
+          // Complex immediates don't have a corresponding *_value() function.
+          // They also only come up for the type's default value, since all
+          // other bits that could be literals also could be computations
+          // involving variables named i (so we need to resolve it).
```
Comment (mppf) [inline compiler/resolution/resolveFunction.cpp]: Shouldn't this case warn or something?

File: compiler/resolution/resolveFunction.cpp [hunk 2]
```diff
@@ -143,6 +149,91 @@ static void resolveFormals(FnSymbol* fn) {
...
+  if (fLibraryPython) {
+    Expr* end = formal->defaultExpr->body.tail;
+
+    if (SymExpr* sym = toSymExpr(end)) {
+      VarSymbol* var = toVarSymbol(sym->symbol());
+
+      // Might be an ArgSymbol instead of a VarSymbol
+      if (var && var->isImmediate()) {
+        Immediate* imm = var->immediate;
+        std::stringstream ss;
```
Comment (mppf) [inline compiler/resolution/resolveFunction.cpp]: I find it a bit odd that you needed a std::stringstream as it's just converting int to string etc... I'd think there'd be a better method to convert an immediate to a string. Could you make a GenRet from the immediate? Or do immediates have some other method? Anyway this code should probably be factored out into an immediate support file.

**Comment by lydia-duncan** [comment]
@mppf - was this what you had in mind?

**Comment by lydia-duncan** [comment]
I added your comment on the default value functions to the issue I made for the more complicated case

**Comment by mppf** [comment]
Yes that's an improvement. My next question - why does it use `std::string` rather than `astr` like the compiler code usually does?

**Comment by lydia-duncan** [comment]
> why does it use `std::string` rather than `astr` like the compiler code usually does?

Because I keep looking up C++ solutions for the problems I'm trying to solve and that leads me to stringstreams.  I also continually forget that we have `astr` lying around in the compiler.

**Comment by mppf** [comment]
I'm not really objecting, it just seemed odd. In some cases there are arguments for `std::string` instead of `astr` - namely when the string is short-lived.

**Comment by lydia-duncan** [comment]
I'm okay with changing it if you think this isn't one of those cases.  I've been thinking of `const char *` and `std::string` as really a language decision (viewing our uses of `const char*` as a sign that the compiler is using outdated styles instead of more modern C++) and so gravitate to `std::string` when given the choice.  It's entirely possible that my thinking there is an oversimplification.

I'm not sure what it will take for me to remember `astr` exists, I'm annoyed at myself for not having it ingrained yet like I do other things in the compiler.  Maybe because I've filed it under "things we implement because we haven't migrated the compiler to a c++11 world", I'm not sure

**Comment by mppf** [comment]
A reasonable viewpoint would be that `astr` is just for "Identifiers" and since you're not making names of identifiers you should use `std::string`.

**Pull request merged**